phpBB/phpbb/textformatter/s9e/mention_helper.php

```diff
@@ -0,0 +1,277 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license       GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\textformatter\s9e;
+
+use s9e\TextFormatter\Utils as TextFormatterUtils;
+
+class mention_helper
+{
+	/**
+	 * @var \phpbb\db\driver\driver_interface
+	 */
+	protected $db;
+
+	/**
+	 * @var \phpbb\auth\auth
+	 */
+	protected $auth;
+
+	/**
+	 * @var \phpbb\user
+	 */
+	protected $user;
+
+	/**
+	 * @var string Base URL for a user profile link, uses {ID} as placeholder
+	 */
+	protected $user_profile_url;
+
+	/**
+	 * @var string Base URL for a group profile link, uses {ID} as placeholder
+	 */
+	protected $group_profile_url;
+
+	/**
+	 * @var array Array of users' and groups' colours for each cached ID
+	 */
+	protected $cached_colours = [];
+
+	/**
+	 * @var array Array of group IDs allowed to be mentioned by current user
+	 */
+	protected $mentionable_groups = null;
+
+	/**
+	 * Constructor
+	 *
+	 * @param \phpbb\db\driver\driver_interface $db
+	 * @param \phpbb\auth\auth                  $auth
+	 * @param \phpbb\user                       $user
+	 * @param string                            $root_path
+	 * @param string                            $php_ext
+	 */
+	public function __construct($db, $auth, $user, $root_path, $php_ext)
+	{
+		$this->db = $db;
+		$this->auth = $auth;
+		$this->user = $user;
+		$this->user_profile_url = append_sid($root_path . 'memberlist.' . $php_ext, 'mode=viewprofile&u={ID}', false);
+		$this->group_profile_url = append_sid($root_path . 'memberlist.' . $php_ext, 'mode=group&g={ID}', false);
+	}
+
+	/**
+	 * Returns SQL query data for colour SELECT request
+	 *
+	 * @param string $type Name type ('u' for users, 'g' for groups)
+	 * @return array Array of SQL SELECT query data for extracting colours for names
+	 */
+	protected function get_colours_sql($type)
+	{
+		switch ($type)
+		{
+			default:
+			case 'u':
+				return [
+					'SELECT' => 'u.user_colour as colour, u.user_id as id',
+					'FROM'   => [
+						USERS_TABLE => 'u',
+					],
+					'WHERE'  => $this->db->sql_in_set('u.user_type', [USER_NORMAL, USER_FOUNDER]),
+				];
+			case 'g':
+				return [
+					'SELECT' => 'g.group_colour as colour, g.group_id as id',
+					'FROM'   => [
+						GROUPS_TABLE => 'g',
+					],
+				];
+		}
+	}
+
+	/**
+	 * Caches colours of users and groups
+	 */
+	protected function cache_colours()
+	{
+		if (count($this->cached_colours) > 0)
+		{
+			return;
+		}
+
+		$types = ['u', 'g'];
+
+		foreach ($types as $type)
+		{
+			$this->cached_colours[$type] = [];
+
+			$query = $this->db->sql_build_query('SELECT', $this->get_colours_sql($type));
```

Comment:
this is gonna return a lot of data

Reply:
yeah, probably we just shouldn't render the user color, just do a simple CSS styling as gooof suggested above.

Reply:
Yes, but the database is queried only once in 5 minutes.
Do you have suggestions for changing that?

Reply:
I agree with @CHItA and @gooof lets not use the user color.
The issue is not that the database won't like it but that migth be a huge amount of data to load and put in cache and it can be noticed and having one crying user every 5 minutes is not really acceptable. But even if it was having that much data in cache (and so in memory) and not that good.